### PR TITLE
fix(handler): deny access when body.allowed is 'false'

### PR DESCRIPTION
**Pull request by jankapunkt**
## Summary

Implements missing access check when body.allowed is 'false'

## Linked issue(s)

https://github.com/oauthjs/node-oauth2-server/pull/532
#89 

## Involved parts of the project
 
`AuthorizeHandler`

## Added tests?

Integration tests added

## OAuth2 standard

https://datatracker.ietf.org/doc/html/rfc6749.html#section-4.1.1

## Reproduction

Clone, checkout branch, test

### CODE REVIEW

**Comment by FStefanni** [comment]
Hi,

the pr seems a sure improvement.
What make me think, is the use of strings (`'false'`) instead of booleans.
So I am not 100% sure this is correct w.r.t. types, and w.r.t. what could truly arrive.
Nevertheless, we could accept this pr, since it is an improvement, and in case catch possible
typing issues when switching to TypeScript.

Regards.

**Review by jorenvandeweyer — APPROVED** [review]

**Comment by jorenvandeweyer** [comment]
> What make me think, is the use of strings (`'false'`) instead of booleans.

The usage of strings is actually correct since all parameters are passed as a string, both in the query as in the body.

**Comment by jankapunkt** [comment]
@HappyZombies @jwerre can you please second-check and merge if everything is okay?

**Comment by jankapunkt** [comment]
@HappyZombies @jwerre please one of you need to 2nd check, since it would make no sense if I review my own PR ;-) If you also see no issues you can merge immediately

**Comment by jankapunkt** [comment]
@Uzlopak would you mind taking a look at this?

File: lib/handlers/authorize-handler.js [hunk 1]
```diff
@@ -77,7 +77,12 @@ AuthorizeHandler.prototype.handle = function(request, response) {
     throw new InvalidArgumentError('Invalid argument: `response` must be an instance of Response');
   }
 
-  if ('false' === request.query.allowed) {
+  const notAllowed = [
+    request.query.allowed,
+    request.body.allowed
+  ].some(allowed => 'false' === allowed);
+
+  if (notAllowed) {
```
Comment (Uzlopak) [inline lib/handlers/authorize-handler.js:85]: Nah. It should be something like 

if (request.query.allowed === 'false' || request.body.allowed === 'false')

Also why is there no check for request.body.allowed === false? Does formData not allow boolean values?

Also: is it safe? Doesn't OAUth2 server add the allowed url fragment when the authorization finished. So actually you would get a redirect which is a Get and not a post. Not that we accidently make it possible to bypass something? Seems unlikely.

**Review by Uzlopak — CHANGES_REQUESTED** [review]

File: lib/handlers/authorize-handler.js (hunk 1 above)
Comment (jankapunkt) [reply lib/handlers/authorize-handler.js:85]: Okay then I would update it towards this:

```js
const notAllowed = ['false', false].some(value => request.query.allowed === value || request.body.allowed === value)
```

Comment (jorenvandeweyer) [reply lib/handlers/authorize-handler.js:85]: FormData does indeed not allow booleans. They only allow strings and blobs.

Comment (Uzlopak) [reply lib/handlers/authorize-handler.js:85]: Why do you like `some()`? Also query can not be a boolean value, as query has by definition only string values. 

```javascript
if (
    request.query.allowed === 'false' ||
    request.body.allowed === 'false' ||
    request.body.allowed === false
)
```

Is much easier to read and probably the most performant too. 

Just wanted to also suggest, that we should actually check for the opposite. Like query.allowed !== 'true'. From security point of view an explicit allowed is better.

Comment (Uzlopak) [reply lib/handlers/authorize-handler.js:85]: @jorenvandeweyer 
UrlEncoded FormData, Thanks for looking up. 

So it should be simply:

```javascript
if (
    request.query.allowed === 'false' ||
    request.body.allowed === 'false'
)
```

Comment (jankapunkt) [reply lib/handlers/authorize-handler.js:85]: Ok, will do

**Comment by jankapunkt** [comment]
Updated and merged development, please review again

**Review by Uzlopak — APPROVED** [review]

**Review by jorenvandeweyer — APPROVED** [review]

**Pull request merged**